### PR TITLE
GeneratedPropertyListContainer implementation

### PR DESCRIPTION
**Pull request by shabak**
There is a list data container supporting generated columns. This is a clone of original GeneratedPropertyContainer from Vaadin. The implementation contains copy of some code of inner classes of GeneratedPropertyContainer, mostly because it has a protected access. I use Vaadin Grid a lot in my project, so I had to implement it for MGrid, because I needed the paging feature and generated columns as well. So I'd like to share id with community. Also I'd like to implement next desirable features such as filtering by generated columns values and removing properties dynamically. 

This is an example application based on viritin example: http://shabak.ru:9292/generated-property-list-container-example/

This is a use case: https://github.com/shabak/spring-data-vaadin-crud/blob/master/src/main/java/crud/vaadin/MainUI.java

Thanks.

### CODE REVIEW

**Comment by mstahv** [comment]
Looks like a pretty good start! I'd love to have this as a typed (and much more intuitive) version at the MGrid level as well. Something related to this API in MTable:

https://github.com/viritin/viritin/blob/master/src/main/java/org/vaadin/viritin/fields/MTable.java#L532-L542

Having the GeneratedPropertyListContainer might be necessary (or at least easiest way) to implement such a thing to MTable.

BTW. Have you looked into [ComponentRendered add-on](https://vaadin.com/directory#!addon/componentrenderer)  by @datenhahn ? It contains ComponentGrid which is a supercool beast, with exactly the kind of API I'd like to use. I have only tried with components, but I'd be surprised if it don't support plain strings as well.

**Comment by datenhahn** [comment]
sorry, my implementation currently only supports Components, as it was built for that specific use case and I felt in my case the tradeoff for in comparison to a more general class was justified. I took away some power from the user, but he can do less wrong and ends up with a cleaner interface.

You can have a look here for the api in action:
https://github.com/datenhahn/componentrenderer/blob/master/componentrenderer-demo/src/main/java/de/datenhahn/vaadin/componentrenderer/demo/ClassicGridWithDecoratorTab.java#L64

But I will have a look at it and see if I can adapt it cleanly to MGrid (don't know when though).

@shabak
Regarding filtering, the basic problem with generated columns is that they are generated on demand for the next bunch of rows when requesting them. In the case of the grid 100 which is the size of the scroll buffer. I think also the table did not support filtering on generated columns. But if the generated column hides a column of its backing container it still is possible to filter the backing container which will result in most cases in the desired behaviour.

I actually built some sample code for some guy who wanted to filter generated Component Columns. Imagine you store food in text-form in the column (e.g. Hamburger) but you want to display an image of a hamburger in the column. Now you filter on the text, enum, whatever row (in our example contains the word "Hamburger")which lies behind the generated column. In my example you would have to type "Hamburger" in the filter field to only let pass the filter images of a hamburger. Of course you could easily adapt that so you have a drop-down or something to choose the filter by the same image as used in the column (of course the filter would still get text "Hamburger" as input, but the user would not see that).

https://gist.github.com/datenhahn/d9ecad5eba14eba1d326#file-componentgridtab-java-L74-L96

In the readme to the component renderer I tried to explain how the rendering in Grid works and where my component-renderer kicks in. In the blue boxes (the containers) on the right side you see how the generated property container hides e.g. a bean item container.

https://github.com/datenhahn/componentrenderer#the-complete-rendering-lifecycle-of-the-component-renderer

TL;DR: filter the BeanItemContainer-Column lying behind the Generated-Column

For me the most annoying thing always was juggling the two containers. For one task you need the bean item container, for the other the generated property container.

**Comment by shabak** [comment]
@datenhahn Thanks for the great explanation of the problem. I agree that the implementation of the filtering for generated columns has such a problem. Actually I prefer server-side filtering techniques, so in my mind it is only way to avoid problems you explain. 
BTW, ComponentRenderer is a great addon! It has many interested details. My favourite is expanded areas for each row. It can be used for the rich editors of entities, and avoid separated pages and popups for entity editors. 

@mstahv Do you like the idea that generated columns are read-only and non-filterable? Thanks

**Comment by datenhahn** [comment]
Just saw that the GeneratedPropertyListContainer + ListContainer work a bit different than the default containers, so my comments probably don't apply anyway.

Regarding generating values with a lambda I played around a bit and attached it as patches for your (shabak) viritin version and the MainUI of your sample. The examples PropertyValueGenerator was to complex for the lambda demonstration, so I stripped it down to one line for demonstration's sake (with much limited functionality of course). One problem I have is, that you still have to pass around lots of classes (model + presentation) so the compiler has access to the types. Maybe that can be cleaned up further somehow. In the current version I think it is not too useful, but perhaps it helps as a first step to a better solution.

[LambdaPropertyValueGenerator.zip](https://github.com/viritin/viritin/files/226860/LambdaPropertyValueGenerator.zip)

**Comment by shabak** [comment]
I've updated the pull request, so as to implement the container at the MGrid level too. 

This is an example application based on viritin example: http://shabak.ru:9292/generated-property-list-container-example/ _(this time it is working)_. 

This is an updated example, has two cases of using, MTables  https://github.com/shabak/spring-data-vaadin-crud/blob/master/src/main/java/crud/vaadin/MainUI.java

**Comment by mstahv** [comment]
Good job, this is starting to look like a really nice improvement. I'd still love to see the generator as such that you only need single lambda expression for converting from T->String. The implementation could e.g. just overload the current version with a one that takes the simple lambda capable type, which is then "piggy bagged" into the current implementation.

Let me know if you like to continue with this. Otherwise when I have next time a moment to spend on Viritin, I'll pull your change in and fine tune it like above.

**Comment by shabak** [comment]
The pull request is updated according the requirements. The lambda support implementation is based on @datenhahn ideas and example (thanks a lot, wouldn't you mind I mentioned you as an author of the lambda generator: https://github.com/viritin/viritin/pull/175/commits/220071458317653b22e9eb52ed4f510fa1ee5513#diff-ef488241e9eda7a1a6f968c6ee567803R16). 

The example app and sources of the example app are updated (see links above).

Thanks

**Comment by mstahv** [comment]
Excellent!

Is there often some other "presentation types" than String.class, especially currently as the Grid don't support components? I guess you could use other "presentation types" with some "renderers", but should String still be a default (with an overloaded alternative API)? That could still simplify the usage a bit.

If you'd still like to help a bit, you could move your the test/example code to src/test/java/.../it/.

**Comment by shabak** [comment]
I'd like to remove P from LambdaPropertyValueGenerator, mostly because at the end of transformations it is always String. So it will simplify the usage.

**Comment by datenhahn** [comment]
Additionally the beantype should be extractable from the listcontainer, just didn't know how do get it from the dynathingy. I think besides string there is number, image and date

**Comment by mstahv** [comment]
Maybe it is best to add both versions, the simpler without type (defaults to String) and a one with the return type parameter.

**Comment by shabak** [comment]
https://github.com/shabak/spring-data-vaadin-crud/blob/master/src/main/java/crud/vaadin/MainUI.java
http://shabak.ru:9292/generated-property-list-container-example/

**Comment by mstahv** [comment]
Can't thing how this could be better at this point :-) I'll play a bit with this soon, merge it and push out a new release

**Comment by shabak** [comment]
glad to read!

**Comment by mstahv** [comment]
@shabak, just noticed that some classes contained some Copyright notices. Is it ok to remove those? All stuff in this project should be apache 2 licensed.

**Comment by mstahv** [comment]
Otherwise I'll need to remove them and do a clean room implementation :-)

**Comment by shabak** [comment]
okay, agree
